docs/proposals/PROPOSAL_B.md

```diff
@@ -0,0 +1,287 @@
+            "cpu": ["AuthenticAMD"]
+          }
+        }
+      }
+    },
+    "cpuIntel": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "cpu": ["GenuineIntel"]
+          }
+        },
+        "kernelCmdline": {
+          "data": {
+            "intel_iommu": "on"
+          }
+        }
+      }
+    },
+    "gpu": {
+      "refs": {
+        "oneOf": [
+          "gpuNVIDIA",
+          "gpuIntel"
+        ]
+      }
+    },
+    "gpuNVIDIA": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "pci": ["0380:10de"]
+          }
+        },
+        "kernelConfiguration": {
+          "data": {
+            "CONFIG_HOTPLUG_PCI_PCIE": "y",
+            "CONFIG_MODULES": "y",
+            "CONFIG_MODULE_UNLOAD": "y",
+            "CONFIG_PCI_MMCONFIG": "y",
+            "CONFIG_DRM_NOUVEAU": "n"
+          }
+        }
+      }
+    },
+    "gpuIntel": {
+      "plugins": {
+        "hardware": {
+          "data": {
+            "pci": ["0380:8086"]
+          }
+        },
+        "kernelConfiguration": {
+          "data": {
+            "CONFIG_DRM": "y",
+            "CONFIG_DRM_I915": "y",
+            "CONFIG_DRM_I915_USERPTR": "y",
+            "CONFIG_DRM_I915_GVT": "m",
+            "CONFIG_DRM_I915_GVT_KVMGT": "m"
+          }
+        },
+        "kernelModules": {
+          "data": {
+            "kvmgt": {
+              "enable_gvt": "1"
+            }
+          }
+        }
+      }
+    }
+  }
+}
+```
+
+</details>
+
+### Namespace and Plugin
+
+!["Namespace and Plugin](img/B_ns-plugin.png "Namespace and Plugin")
+
+Namespaces and plugins are core functionality to evaluate if a host is compatible with a container.
+They can look up existing components on a host or just serve as a data source to get additional information about specific resources (like CPU features).
+Although in each case a plugin has to return information about success or failure.
+
+The design allows you to build plugins on top of other plugins.
+For instance an Nvidia GPU plugin could include logic of kernelModules and kernelConfiguration to discover if appropriate kernel modules are loaded or disabled.
+
+#### Plugins Categorization
+
+Plugins are developed by:
+* OCI Image Compatibility maintainers (atomic plugins)
+* Community (custom plugins)
+
+Atomic plugins serve as the basis for external plugins provided by a community with expertise in specific areas.
+They are divided into the following areas (similarly like in [Node Feature Discovery](https://kubernetes-sigs.github.io/node-feature-discovery/v0.15/usage/features.html#table-of-contents)):
+* CPU
+* Kernel
+* Memory
+* Network
+* PCI
+* USB
+* Storage
+* System
+
+*Note: We could import part of the NFD project or export the common part and contribute.* 
+
+#### Distribution of Plugins
+
+* Pulled over HTTP - from the address specified in the schema `source` field, pulled to `/opt/oci/image-compatibility/<artifact-digest/plugins`. 
```

Comment:
Thats an attack vector like javascript. For security plugins must be specified eithe in a central repo or a vendor verification mechnism such as GPG. We will run code on the host.

Reply:
Yeah I didn't add a section about security yet. However, you're right there would have to be a vendor verification mechanism.

Reply:
Is the plugin done as a container or is it supposed to run on the host? HTTPS would be preferred, though.

Reply:
Supposed to run on the host. Right, that should be HTTPS.

I think I've opened a pandora box from security point of view. The verification logic for plugins that can be build by anybody can be too complex for this project. Probably I have to think about something much simpler and change the distribution and plugins execution idea to something else.

Reply:
One way to somplify that is to simply define the ABI and language to follow these two rules:
1. It is only possible to output a value to the calling program
2. The language is limited to an interpreted purpose built language like AWK.

Reply:
Yeah that could be a way for binary compatibility, although, it's not only about that. I'm going to change this plugin idea to something less open.